rfcs/0024-scaling-epic.md

```diff
@@ -0,0 +1,144 @@
+
+## Actionable Dependencies
+[actionable-dependencies]: #actionable-dependencies
+
+Actionable dependencies are prioritized on a scale of 1-3, 1 being most important, 3 being least important. All of these have to be done, but some of them have a wider impact or will take more time than others. Each of these actionable items will have their own RFC which is linked back to this RFC before it is closed.
+
+### Optimize Epoch Ledger Synchronization
+##### Priority: 1
+##### Why this Priority?
+This is the primary blocker in scalilng the number of accounts in the ledger.
+##### Why is it important?
+The amount of bandwidth used during epoch ledger synchronization is currently ~O(n^2) with respect to the number of accounts in the ledger. Furthermore, this data is currently sent in a single network request from a single peer. As such, this is the primary factor which will prevent us from scaling the number of accounts in the ledger.
+##### Proposed Fix
+Synchronize the Epoch Ledger using the same Syncable Ledger abstraction we use for synchronizing the Root Snarked Ledger of the Transition Frontier.
+##### Child RFC: &#x1F534;
+
+### Reduce Epoch Ledger Memory Usage
+##### Priority: 2
+##### Why this Priority?
+Less important than optimizing epoch ledger synchronization time/bandwidth usage.
+##### Why is it important?
+Epoch ledgers are currently stored in memory, in their entirety, with no shared memory between the data structures or what is already stored on disk. This means that 2 full copies of the ledger are stored in memory, which puts insane memory requirements on consensus nodes on the network when the ledger is large.
+##### Proposed Fix
+There are multiple potential fixes that need to be weighed and compared for implementation time and resulting improvements. Here are some examples:
+- persistent backwards chained masks (same as above, but on disk with an in-memory cache for relevant staking accounts)
+- content addressed storage system (see @enolan's RFC for details)
+##### Child RFC: &#x1F534;
+
+### Reduce Scan State Memory Usage
+##### Priority: 3
+##### Why this Priority?
+This is important and will cause issues at some point, but it looks like this will show itself as a bottleneck later than most of the other scaling dependencies.
+##### Why is it important?
+At high parameterization, the scan state will use an increadible amount of memory. The memory usage needs to be before we can significantly turn up the scan state's size.
+##### Proposed Fix
+Don't persist completed work in the trees, throw out completed trees from the tree ring as we go. Use content addressed storage to share references to memory across transaction witnesses (which potentially saves a lot of duplicate intermediate hashes from being stored).
```

Comment:
Throwing out completed trees is already implemented

Reply:
Great.